### PR TITLE
Sync to v2.3

### PR DESCRIPTION
**Pull request by diego-G**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
+> This repository is still under development.
+>
+> The majority of tests provided are not updated and are incomplete.
+
 This repository contains the configuration to manage CMTAT using the
 [Foundry](https://book.getfoundry.sh/) suite – If you use Truffle instead of Foundry, please see
 [CMTA/CMTAT-Truffle](https://github.com/CMTA/CMTAT-Truffle).
 
-The CMTAT contracts are included as a [submodule](CMTAT/) of the present repository. The current version used is the version [2.1](https://github.com/CMTA/CMTAT/releases/tag/2.1).
+The CMTAT contracts are included as a [submodule](CMTAT/) of the present repository. The current version used is the version [2.3]([https://github.com/CMTA/CMTAT/releases/tag/2.3-Beta](https://github.com/CMTA/CMTAT/releases/tag/v2.3.0)).
```
Comment (zarifpour) [inline README.md:13]: There is an issue with the hyperlink here.

```suggestion
The CMTAT contracts are included as a [submodule](CMTAT/) of the present repository. The current version used is the version [2.3](https://github.com/CMTA/CMTAT/releases/tag/v2.3.0).
```

File: README.md [hunk 2]
```diff
@@ -1,10 +1,16 @@
 # CMTAT - using the Foundry suite
 
+> Warning:
+>
+> This repository is still under development.
+>
+> The majority of tests provided are not updated and are incomplete.
```
Comment (zarifpour) [inline README.md:7]:
```suggestion
> [!Warning]
> This repository is still under development.
> The majority of tests provided are not updated and are incomplete.
```
> [!Warning]
> This repository is still under development.
> The majority of tests provided are not updated and are incomplete.

OR

> [!Caution]
> This repository is still under development.
> The majority of tests provided are not updated and are incomplete.

File: test/ValidationModule/RuleEngineSet.t.sol [hunk 3]
```diff
@@ -3,29 +3,34 @@ pragma solidity ^0.8.17;
...
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
+        CMTAT_CONTRACT = new CMTAT_STANDALONE(
+            ZERO_ADDRESS,
+            DEFAULT_ADMIN_ADDRESS,
             "CMTA Token",
             "CMTAT",
             "CMTAT_ISIN",
-            "https://cmta.ch");
+            "https://cmta.ch",
+             IRuleEngine(address(0)),
```
Comment (zarifpour) [inline test/ValidationModule/RuleEngineSet.t.sol:21]: extra space

File: test/ValidationModule/CodeListInternal.sol [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 pragma solidity ^0.8.17;
 
-abstract contract CodeList {
+abstract contract CodeListInternal {
```
Comment (zarifpour) [inline test/ValidationModule/CodeListInternal.sol:5]: Should we keep the name `CodeList` and move it into its own director `ValidationModule/Internal`?

File: test/ValidationModule/CodeListInternal.sol [hunk 5]
```diff
@@ -2,7 +2,7 @@
...
```
Comment (zarifpour) [inline test/ValidationModule/CodeListInternal.sol:6]: extra space

File: test/BurnModule.t.sol [hunk 6]
```diff
@@ -25,18 +30,19 @@ contract BurnModuleTest is Test, HelperContract, BurnModule {
...
 
         // Burn 20
         // Assert
         vm.expectEmit(true, true, false, true);
         emit Transfer(ADDRESS1, ZERO_ADDRESS, 20);
-        vm.expectEmit(true, false, false, true);
-        emit Burn(ADDRESS1, 20);
+        // TODO: add reason argument
+        //vm.expectEmit(true, false, false, true);
+        //emit Burn(ADDRESS1, 20);
```
Comment (zarifpour) [inline test/BurnModule.t.sol:41]: Should this be done before merging?

File: test/BurnModule.t.sol [hunk 7]
```diff
@@ -25,18 +30,19 @@ contract BurnModuleTest is Test, HelperContract, BurnModule {
...
 
         // Act
-        vm.prank(ADMIN_ADDRESS);
-        CMTAT_CONTRACT.forceBurn(ADDRESS1, 20);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
+        CMTAT_CONTRACT.forceBurn(ADDRESS1, 20, "");
```
Comment (zarifpour) [inline test/BurnModule.t.sol:46]: Extra space

File: test/BurnModule.t.sol [hunk 8]
```diff
@@ -49,12 +55,16 @@ contract BurnModuleTest is Test, HelperContract, BurnModule {
         // Assert
         vm.expectEmit(true, true, false, true);
         emit Transfer(ADDRESS1, ZERO_ADDRESS, 30);
+        /*
+        // TODO: add reason argument
         vm.expectEmit(true, false, false, true);
         emit Burn(ADDRESS1, 30);
-        vm.prank(ADMIN_ADDRESS);
+        
+        */
```
Comment (zarifpour) [inline test/BurnModule.t.sol:63]: Should this be done before merging?

File: test/BurnModule.t.sol [hunk 9]
```diff
@@ -66,20 +76,22 @@ contract BurnModuleTest is Test, HelperContract, BurnModule {
...
 
         // Assert
         vm.expectEmit(true, true, false, true);
         emit Transfer(ADDRESS1, ZERO_ADDRESS, 20);
+        /*
+        // TODO: add reason argument
         vm.expectEmit(true, false, false, true);
         emit Burn(ADDRESS1, 20);
-
+        */
```
Comment (zarifpour) [inline test/BurnModule.t.sol:91]: Should this be done before merging?

File: test/PauseModule.t.sol [hunk 10]
```diff
@@ -23,19 +29,19 @@ contract PauseModuleTest is Test, HelperContract, PauseModule {
     */
     function testCanBePausedByAdmin() public {
         // Act
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         vm.expectEmit(false, false, false, true);
-        emit Paused(ADMIN_ADDRESS);
+        emit Paused(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.pause();
         // Assert
```
Comment (zarifpour) [inline test/PauseModule.t.sol:36]: space

File: test/PauseModule.t.sol [hunk 11]
```diff
@@ -23,19 +29,19 @@ contract PauseModuleTest is Test, HelperContract, PauseModule {
...
         vm.prank(ADDRESS1);
-        vm.expectRevert(bytes("CMTAT: token transfer while paused"));
+        vm.expectRevert(bytes("CMTAT: transfer rejected by validation module"));
```
Comment (zarifpour) [inline test/PauseModule.t.sol:38]: We can make this message into a variable since it is reused.

File: test/MintModule.t.sol [hunk 12]
```diff
@@ -62,11 +67,11 @@ contract MintModuleTest is Test, HelperContract, MintModule {
 
     function testCanBeMintedByANewMinter() public {
         // Arrange
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.grantRole(MINTER_ROLE, ADDRESS1);
         // Arrange - Assert
```
Comment (zarifpour) [inline test/MintModule.t.sol:72]: space

File: test/HelperContract.sol [hunk 13]
```diff
@@ -2,13 +2,14 @@
 pragma solidity ^0.8.17;
 
 import "forge-std/Test.sol";
-import "CMTAT/CMTAT.sol";
-
+import "CMTAT/CMTAT_STANDALONE.sol";
+import "CMTAT/mocks/RuleEngine/interfaces/IRuleEngine.sol";
```
Comment (zarifpour) [inline test/HelperContract.sol:6]: Are these intentionally unused (excluding CMTAT_STANDALONE)?

File: test/HelperContract.sol [hunk 14]
```diff
@@ -2,13 +2,14 @@
...
 abstract contract HelperContract {
-    CMTAT CMTAT_CONTRACT;
+    uint256 FLAG = 5;
+    CMTAT_STANDALONE CMTAT_CONTRACT;
     address constant ZERO_ADDRESS = address(0);
-    address constant OWNER = address(1);
-    address constant ADMIN_ADDRESS = address(5);
+    address constant DEFAULT_ADMIN_ADDRESS = address(1);
+    //address constant ADMIN_ADDRESS = address(5);
```
Comment (zarifpour) [inline test/HelperContract.sol:12]: - Should we set the `DEFAULT_ADMIN_ADDRESS` to 5?
- Let's remove the comment?

File: test/EnforcementModule.sol [hunk 15]
```diff
@@ -32,63 +37,73 @@ contract EnforcementAuthorizationModule is
...
+        /*
         vm.expectEmit(true, true,false, false);
-        emit Freeze(ADMIN_ADDRESS, ADDRESS1);
-        vm.prank(ADMIN_ADDRESS);
-        CMTAT_CONTRACT.freeze(ADDRESS1);
+        emit Freeze(DEFAULT_ADMIN_ADDRESS, ADDRESS1);
+        */
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
+        CMTAT_CONTRACT.freeze(ADDRESS1, "");
         // Assert
```
Comment (zarifpour) [inline test/EnforcementModule.sol:46]: space

File: test/EnforcementModule.sol [hunk 16]
```diff
@@ -32,63 +37,73 @@ contract EnforcementAuthorizationModule is
         resBool = CMTAT_CONTRACT.frozen(ADDRESS1);
         assertFalse(resBool);
         // Act
+        /*
         vm.expectEmit(true, true,false, false);
-        emit Freeze(ADMIN_ADDRESS, ADDRESS1);
-        vm.prank(ADMIN_ADDRESS);
-        CMTAT_CONTRACT.freeze(ADDRESS1);
+        emit Freeze(DEFAULT_ADMIN_ADDRESS, ADDRESS1);
+        */
```
Comment (zarifpour) [inline test/EnforcementModule.sol:43]: comment?

File: test/EnforcementModule.sol [hunk 17]
```diff
@@ -32,63 +37,73 @@ contract EnforcementAuthorizationModule is
...
     function testEnforcerRoleCanFreezeAddress() public {
         // Arrange
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.grantRole(ENFORCER_ROLE, ADDRESS2);
         // Arrange - Assert
         resBool = CMTAT_CONTRACT.frozen(ADDRESS1);
         assertFalse(resBool);
         // Act
+        // TODO: add reason argument
```
Comment (zarifpour) [inline test/EnforcementModule.sol:59]: Should we resolve these TODOs before merging?

**Review by zarifpour — CHANGES_REQUESTED** [review]

File: test/HelperContract.sol (hunk 13 above)
Comment (mariogutval) [reply test/HelperContract.sol:6]: I think we can remove it

File: test/ValidationModule/RuleEngineTransfer.sol [hunk 18]
```diff
@@ -3,34 +3,39 @@ pragma solidity ^0.8.17;
...
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.mint(ADDRESS1, 31);
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.mint(ADDRESS2, 32);
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
         CMTAT_CONTRACT.mint(ADDRESS3, 33);
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/ValidationModule/RuleEngineTransfer.sol:38]: We can use `startPrank` and `stopPrank` cheatcodes instead?

File: test/AuthorizationModule.t.sol [hunk 19]
```diff
@@ -4,6 +4,7 @@ pragma solidity ^0.8.17;
 import "forge-std/Test.sol";
 import "./HelperContract.sol";
 
+
```
Comment (mariogutval) [inline test/AuthorizationModule.t.sol:7]:
```suggestion
```

File: test/BaseModule.t.sol [hunk 20]
```diff
@@ -274,22 +284,26 @@ contract AllowanceTest is Test, HelperContract, BaseModule, ERC20BaseModule {
 
 contract TransferTest is Test, HelperContract, BaseModule, ERC20BaseModule {
     function setUp() public {
-        vm.prank(OWNER);
-        CMTAT_CONTRACT = new CMTAT(ZERO_ADDRESS, false,
-            OWNER,
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/BaseModule.t.sol:287]: We can use `startPrank` and `stopPrank` here too.

File: test/EnforcementModule.sol [hunk 21]
```diff
@@ -13,14 +13,19 @@ contract EnforcementAuthorizationModule is
     bool resBool;
 
     function setUp() public {
-        vm.prank(ADMIN_ADDRESS);
-        CMTAT_CONTRACT = new CMTAT(ZERO_ADDRESS, false,
-            ADMIN_ADDRESS,
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/EnforcementModule.sol:16]: We can use `startPrank` and `stopPrank` here too.

File: test/MintModule.t.sol [hunk 22]
```diff
@@ -21,12 +26,12 @@ contract MintModuleTest is Test, HelperContract, MintModule {
...
         // Check first balance
-        uint256 res1 = CMTAT_CONTRACT.balanceOf(ADMIN_ADDRESS);
+        uint256 res1 = CMTAT_CONTRACT.balanceOf(DEFAULT_ADMIN_ADDRESS);
         assertEq(res1, 0);
 
         // Act
         // Issue 20 and check balances and total supply
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/MintModule.t.sol:34]: We can use `startPrank` and `stopPrank` here too.

File: test/MintModule.t.sol [hunk 23]
```diff
@@ -62,11 +67,11 @@ contract MintModuleTest is Test, HelperContract, MintModule {
 
     function testCanBeMintedByANewMinter() public {
         // Arrange
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/MintModule.t.sol:70]: We can use `startPrank` and `stopPrank` here too.

File: test/PauseModule.t.sol [hunk 24]
```diff
@@ -5,16 +5,22 @@ import "forge-std/Test.sol";
 import "./HelperContract.sol";
 
 contract PauseModuleTest is Test, HelperContract, PauseModule {
+    
     function setUp() public {
-        vm.prank(ADMIN_ADDRESS);
-        CMTAT_CONTRACT = new CMTAT(ZERO_ADDRESS, false,
-            ADMIN_ADDRESS,
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/PauseModule.t.sol:10]: We can use `startPrank` and `stopPrank` here too.

File: test/PauseModule.t.sol [hunk 25]
```diff
@@ -70,15 +76,15 @@ contract PauseModuleTest is Test, HelperContract, PauseModule {
 
     function testCanBeUnpausedByAdmin() public {
         // Arrange
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/PauseModule.t.sol:79]: We can use `startPrank` and `stopPrank` here too.

File: test/PauseModule.t.sol [hunk 26]
```diff
@@ -89,9 +95,9 @@ contract PauseModuleTest is Test, HelperContract, PauseModule {
 
     function testCanBeUnpausedByANewPauser() public {
         // Arrange
-        vm.prank(ADMIN_ADDRESS);
+        vm.prank(DEFAULT_ADMIN_ADDRESS);
```
Comment (mariogutval) [inline test/PauseModule.t.sol:98]: We can use `startPrank` and `stopPrank` here too.

**Review by mariogutval — CHANGES_REQUESTED** [review]
Only minor comments. In general, review the use of `prank` vs. the combination of `startPrank` and `stopPrank` to avoid repetition.

**Pull request closed without merging**